ci/scripts/cpp_build.sh

```diff
@@ -189,6 +189,9 @@ cmake \
 export CMAKE_BUILD_PARALLEL_LEVEL=${CMAKE_BUILD_PARALLEL_LEVEL:-$[${n_jobs} + 1]}
 time cmake --build . --target install
 
+# Save disk space by removing large temporary build products
+find . -name "*.o" -delete
```

Comment:
Wouldn't that also clean the executables?

Reply:
Indeed, it removes all compiled executables, which makes it pointless here.

Reply:
Ah, does our integration test use non-installed executables?
I think that we should install these executables.

Reply:
Where would we install them? We run all C++ tests from `/build/cpp`.

Reply:
I will check.

Reply:
https://github.com/apache/arrow/blob/3cb4481386876c8004b2ab0b1bae58b658ef6ef0/docker-compose.yml#L1705-L1706

Reply:
Also, the integration utility is not installed:
```console
# ls /opt/conda/envs/arrow/bin/arrow*
/opt/conda/envs/arrow/bin/arrow-file-to-stream  /opt/conda/envs/arrow/bin/arrow-stream-to-file
# ls /build/cpp/debug/arrow*
/build/cpp/debug/arrow-file-to-stream  /build/cpp/debug/arrow-json-integration-test  /build/cpp/debug/arrow-stream-to-file
```

And regardless, `cpp_build.sh` is used by many other builds than the integration build.

Reply:
I imagined the following:

```diff
diff --git a/ci/scripts/cpp_build.sh b/ci/scripts/cpp_build.sh
index 5a89fafc6..f9cbf4bf3 100755
--- a/ci/scripts/cpp_build.sh
+++ b/ci/scripts/cpp_build.sh
@@ -192,7 +192,7 @@ time cmake --build . --target install
 popd
 
 if [ -x "$(command -v ldconfig)" ]; then
-  ldconfig ${ARROW_HOME}/${CMAKE_INSTALL_LIBDIR:-lib}
+  ldconfig ${CMAKE_INSTALL_PREFIX:-${ARROW_HOME}}/${CMAKE_INSTALL_LIBDIR:-lib}
 fi
 
 if [ "${ARROW_USE_CCACHE}" == "ON" ]; then
diff --git a/cpp/src/arrow/flight/integration_tests/CMakeLists.txt b/cpp/src/arrow/flight/integration_tests/CMakeLists.txt
index 98a7a2a7a..a453f858e 100644
--- a/cpp/src/arrow/flight/integration_tests/CMakeLists.txt
+++ b/cpp/src/arrow/flight/integration_tests/CMakeLists.txt
@@ -28,18 +28,21 @@ list(APPEND
      ${GFLAGS_LIBRARIES}
      ${ARROW_GTEST_GTEST})
 
-add_executable(flight-test-integration-server test_integration_server.cc
+add_executable(arrow-cpp-flight-test-integration-server test_integration_server.cc
                                               test_integration.cc)
-target_link_libraries(flight-test-integration-server
+target_link_libraries(arrow-cpp-flight-test-integration-server
                       ${ARROW_FLIGHT_INTEGRATION_TEST_LINK_LIBS})
 
-add_executable(flight-test-integration-client test_integration_client.cc
+add_executable(arrow-cpp-flight-test-integration-client test_integration_client.cc
                                               test_integration.cc)
-target_link_libraries(flight-test-integration-client
+target_link_libraries(arrow-cpp-flight-test-integration-client
                       ${ARROW_FLIGHT_INTEGRATION_TEST_LINK_LIBS})
 
-add_dependencies(arrow-integration flight-test-integration-client
-                 flight-test-integration-server)
+install(TARGETS arrow-cpp-flight-test-integration-server arrow-cpp-flight-test-integration-client ${INSTALL_IS_OPTIONAL}
+          DESTINATION ${CMAKE_INSTALL_BINDIR})
+
+add_dependencies(arrow-integration arrow-cpp-flight-test-integration-client
+                 arrow-cpp-flight-test-integration-server)
 
 if(ARROW_BUILD_TESTS)
   add_arrow_test(flight_integration_test
diff --git a/cpp/src/arrow/testing/CMakeLists.txt b/cpp/src/arrow/testing/CMakeLists.txt
index 19c2a83ef..a3720abfd 100644
--- a/cpp/src/arrow/testing/CMakeLists.txt
+++ b/cpp/src/arrow/testing/CMakeLists.txt
@@ -26,12 +26,14 @@ endif()
 # - a self-(unit)test for the C++ side of integration testing
 if(ARROW_BUILD_TESTS)
   add_arrow_test(json_integration_test EXTRA_LINK_LIBS ${GFLAGS_LIBRARIES})
-  add_dependencies(arrow-integration arrow-json-integration-test)
-elseif(ARROW_BUILD_INTEGRATION)
-  add_executable(arrow-json-integration-test json_integration_test.cc)
-  target_link_libraries(arrow-json-integration-test ${ARROW_TEST_LINK_LIBS}
-                        ${GFLAGS_LIBRARIES} ${ARROW_GTEST_GTEST})
-
-  add_dependencies(arrow-json-integration-test arrow arrow_testing)
-  add_dependencies(arrow-integration arrow-json-integration-test)
+endif()
+if(ARROW_BUILD_INTEGRATION)
+  add_executable(arrow-cpp-json-integration-test json_integration_test.cc)
+  target_link_libraries(arrow-cpp-json-integration-test ${ARROW_TEST_LINK_LIBS}
+                        ${GFLAGS_LIBRARIES} ${ARROW_GTEST_GTEST})
+  install(TARGETS arrow-cpp-json-integration-test ${INSTALL_IS_OPTIONAL}
+            DESTINATION ${CMAKE_INSTALL_BINDIR})
+
+  add_dependencies(arrow-cpp-json-integration-test arrow arrow_testing)
+  add_dependencies(arrow-integration arrow-cpp-json-integration-test)
 endif()
diff --git a/dev/archery/archery/integration/tester_cpp.py b/dev/archery/archery/integration/tester_cpp.py
index 52cc565dc..362f59d62 100644
--- a/dev/archery/archery/integration/tester_cpp.py
+++ b/dev/archery/archery/integration/tester_cpp.py
@@ -24,17 +24,13 @@ from .util import run_cmd, log
 from ..utils.source import ARROW_ROOT_DEFAULT
 
 
-_EXE_PATH = os.environ.get(
-    "ARROW_CPP_EXE_PATH", os.path.join(ARROW_ROOT_DEFAULT, "cpp/build/debug")
-)
-_INTEGRATION_EXE = os.path.join(_EXE_PATH, "arrow-json-integration-test")
-_STREAM_TO_FILE = os.path.join(_EXE_PATH, "arrow-stream-to-file")
-_FILE_TO_STREAM = os.path.join(_EXE_PATH, "arrow-file-to-stream")
+_INTEGRATION_EXE = "arrow-cpp-json-integration-test"
+_STREAM_TO_FILE = "arrow-stream-to-file"
+_FILE_TO_STREAM = "arrow-file-to-stream"
 
-_FLIGHT_SERVER_CMD = [os.path.join(
-    _EXE_PATH, "flight-test-integration-server")]
+_FLIGHT_SERVER_CMD = ["arrow-cpp-flight-test-integration-server"]
 _FLIGHT_CLIENT_CMD = [
-    os.path.join(_EXE_PATH, "flight-test-integration-client"),
+    "arrow-cpp-flight-test-integration-client",
     "-host",
     "localhost",
 ]
diff --git a/docker-compose.yml b/docker-compose.yml
index c193a7d11..8338b23a8 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -1702,8 +1702,6 @@ services:
     volumes: *conda-volumes
     environment:
       <<: [*common, *ccache]
-      # tell archery where the arrow binaries are located
-      ARROW_CPP_EXE_PATH: /build/cpp/debug
       ARCHERY_INTEGRATION_WITH_RUST: 0
     command:
       ["/arrow/ci/scripts/rust_build.sh /arrow /build &&
```

But it didn't work because depending `libgtest.so` isn't installed.

Reply:
That's a lot of changes just so that we can use the clean target...

Reply:
I'll do it later: #37505
I think that using executables in build directory is a bit ad-hoc.